src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,181 @@
+import { Env } from '../config';
+import { InternalHttpServiceSetup, KibanaRequest, KibanaResponseFactory } from '../http';
+import { IUiSettingsClient } from '../ui_settings';
+import { PluginsServiceSetup, DiscoveredPlugin } from '../plugins';
+import { createCspRuleString } from '../csp';
+import { mergeVariables } from './merge_variables';
+import { IApplicationContextProvider, PluginVariables, UserVariables } from './types';
+
+interface Params {
+  request: KibanaRequest;
+  response: KibanaResponseFactory;
+  env: Env;
+  http: InternalHttpServiceSetup;
+  uiSettings: IUiSettingsClient;
+  plugins: PluginsServiceSetup;
+}
+
+interface ConfiguredPlugin {
+  id: string;
+  plugin: DiscoveredPlugin;
+  config: unknown;
+}
+
+const template = compileFile(join(__dirname, 'views/ui_app.pug'));
+// TODO: Remove once https://github.com/elastic/kibana/pull/51438 lands:
+const defaultCapabilities: Capabilities = {
+  navLinks: {},
+  management: {},
+  catalogue: {},
+};
+
+export class ApplicationContextProvider implements IApplicationContextProvider {
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: IUiSettingsClient;
+  private readonly plugins!: PluginsServiceSetup;
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
+
+  private async getUiPlugins(): Promise<ConfiguredPlugin[]> {
+    return Promise.all(
+      [...this.plugins.uiPlugins.public].map(async ([id, plugin]) => {
+        const config = await this.plugins.uiPluginConfigs
+          .get(id)
+          ?.pipe(take(1))
+          .toPromise();
+
+        return { id, plugin, config: config || {} };
+      })
+    );
+  }
+
+  private async getMetadata(
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const [uiPlugins, uiSettings, userSettings] = await Promise.all<
+      ConfiguredPlugin[],
+      PluginVariables,
+      UserVariables
+    >([
+      this.getUiPlugins(),
+      this.uiSettings.getAll(),
+      includeUserProvidedConfig ? this.uiSettings.getUserProvided() : Promise.resolve({}),
+    ]);
+
+    return {
+      uiPlugins,
+      uiSettings,
+      userSettings,
+      vars: mergeVariables(
+        uiPlugins.map(({ config }) => config),
+        injectedOverrides
+      ),
+    };
+  }
+
+  private async getVariables(
+    pluginId: string,
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const { uiPlugins, uiSettings, userSettings, vars } = await this.getMetadata(
+      injectedOverrides,
+      includeUserProvidedConfig
+    );
+    const basePath = this.http.basePath.get(this.request);
+    // TODO: Use full implementation once https://github.com/elastic/kibana/pull/51438 lands:
+    const capabilities = defaultCapabilities;
+
+    return {
+      uiPublicUrl: `${basePath}/ui`,
```

Comment:
`strictCsp` is missed https://github.com/restrry/kibana/blob/2f87fd624aab8c3c5fbe65094ae2a72e38dcb51e/src/legacy/ui/ui_render/ui_render_mixin.js#L250

Reply:
Yeah, I moved this all to the `injectedMetadata.csp` property, but since it sounds like we want legacy to use this, then I will add it back.